### PR TITLE
Inheritance - Initial impl. of relational query support:

### PR DESCRIPTION
**Pull request by anpete**
- EntityMaterializerSource can now return expressions to support tree composition.
- Added support for OfType result operator.
- Introduced ExpressionStringBuilder for complete control over query cache key generation.
- Added MaterializerFactory to relational, relational materializers need to take discriminators into account.
- Added Discriminator annotation metadata APIs to relational.

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs [hunk 1]
```diff
@@ -103,8 +95,36 @@ Expression readValueExpression
...
+            [CanBeNull] int[] indexMap = null)
         {
+            Check.NotNull(entityType, nameof(entityType));
+            Check.NotNull(valueReaderExpression, nameof(valueReaderExpression));
+
+            var materializer = entityType as IEntityMaterializer;
+
+            if (materializer != null)
+            {
+                ((Func<IValueReader, object>)materializer.CreateEntity).GetMethodInfo();
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs]: Remove line 110

File: src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs [hunk 2]
```diff
@@ -125,14 +145,13 @@ let targetMember
                 select Expression.Assign(
                     targetMember,
                     CreateReadValueExpression(
-                        _readerParameter, targetMember.Type, mapping.Item1.Index)));
+                        valueReaderExpression,
+                        targetMember.Type,
+                        indexMap?[mapping.Item1.Index] ?? mapping.Item1.Index)));
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs]: Reformat

File: src/EntityFramework.Core/Query/ExpressionTreeVisitors/ExpressionStringBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,859 @@
...
+            }
+        }
+
+        private int GetLabelId(LabelTarget label)
+        {
+            if (_ids == null)
+            {
+                _ids = new Dictionary<object, int>();
+                AddLabel(label);
+                return 0;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Query/ExpressionTreeVisitors/ExpressionStringBuilder.cs]: Remove lines 66 and 67

File: src/EntityFramework.Core/Query/ExpressionTreeVisitors/ExpressionStringBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,859 @@
...
+            }
+        }
+
+        private int GetParamId(ParameterExpression p)
+        {
+            if (_ids == null)
+            {
+                _ids = new Dictionary<object, int>();
+                AddParam(p);
+                return 0;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Query/ExpressionTreeVisitors/ExpressionStringBuilder.cs]: Remove lines 100 and 101

File: src/EntityFramework.Core/Query/ExpressionTreeVisitors/ExpressionStringBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,859 @@
...
+        protected override Expression VisitParameter(ParameterExpression node)
+        {
+            if (node.IsByRef)
+            {
+                Out("ref ");
+            }
+            var name = node.Name;
+            if (string.IsNullOrEmpty(name))
+            {
+                Out("Param_" + GetParamId(node));
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Query/ExpressionTreeVisitors/ExpressionStringBuilder.cs]: Consider just "P" as the suffix

File: src/EntityFramework.Relational/Metadata/ReadOnlyRelationalEntityTypeExtensions.cs [hunk 6]
```diff
@@ -21,19 +23,19 @@ public ReadOnlyRelationalEntityTypeExtensions([NotNull] IEntityType entityType)
...
 
-        protected virtual IEntityType EntityType
-        {
-            get { return _entityType; }
-        }
+        public virtual IProperty DiscriminatorProperty
+            => _entityType.RootType
+                .GetProperty(
+                    _entityType.RootType
+                        .GetAnnotation(DiscriminatorPropertyAnnotation).Value);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Relational/Metadata/ReadOnlyRelationalEntityTypeExtensions.cs]: Use indexer

**Comment by AndriySvyryd** [comment]
:shipit:

File: src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs (hunk 1 above)
Comment (anpete) [inline src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs]: Thanks.

File: src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs (hunk 2 above)
Comment (anpete) [inline src/EntityFramework.Core/Metadata/EntityMaterializerSource.cs]: Done.

File: src/EntityFramework.Relational/Metadata/ReadOnlyRelationalEntityTypeExtensions.cs (hunk 6 above)
Comment (anpete) [inline src/EntityFramework.Relational/Metadata/ReadOnlyRelationalEntityTypeExtensions.cs]: It should throw here.

**Pull request merged**